common/djangoapps/student/models.py

```diff
@@ -1552,11 +1553,163 @@ def send_signal_full(cls, event, user=user, mode=mode, course_id=None, cost=None
+                'assignment_type': '',
+                'link': '',
+                'assignment_count': -1,
+                'due_date': '',
+                'due_time': ''
+            },
+            {
+                'id': 2,
+                'date': '',
+                'title': '',
+                'assignment_type': '',
+                'link': '',
+                'assignment_count': -1,
+                'due_date': '',
+                'due_time': ''
+            },
+            {
+                'id': 3,
+                'date': '',
+                'title': '',
+                'assignment_type': '',
+                'link': '',
+                'assignment_count': -1,
+                'due_date': '',
+                'due_time': ''
+            }
+        ]
+        today = datetime.now(UTC).date()
+
+        for block in blocks:
+            block_date = datetime.strptime(block['date'][:19], '%Y-%m-%dT%H:%M:%S')
+            block_date = block_date.replace(tzinfo=UTC).date()
+
+            if block_date < today:
+                course_date_list[0]['date'] = block_date
+            if block_date == today:
+                course_date_list[1]['date'] = block_date
+            if block_date > today and course_date_list[2]['date'] == '':
+                course_date_list[2]['date'] = block_date
+
+        if course_date_list[1]['date'] == '':
+            course_date_list[1]['date'] = today
+            course_date_list[1]['due_date'] = today.strftime("%a, %b %d, %Y")
+        if course_date_list[0]['date'] == '':
+            course_date_list[0]['date'] = course_date_list[1]['date']
+        if course_date_list[2]['date'] == '':
+            course_date_list[2]['date'] = course_date_list[1]['date']
+        for block in blocks:
+            block_date = datetime.strptime(block.get('date')[:19], '%Y-%m-%dT%H:%M:%S')
+            block_date = block_date.replace(tzinfo=UTC)
+            timezone_offset = get_time_zone_offset(user_timezone, block_date)
+            block_date = block_date.astimezone(user_timezone)
+
+            if block_date.date() == course_date_list[0]['date']:
+                if course_date_list[0]['assignment_count'] == -1:
+                    course_date_list[0].update({
+                        'title': block['title'] or '',
+                        'assignment_type': block['assignment_type'] or '',
+                        'link': block['link'] or '',
+                        'assignment_count': (course_date_list[0]['assignment_count'] + 1 if block['assignment_type']
+                                             else course_date_list[0]['assignment_count']),
+                        'due_date': block_date.strftime("%a, %b %d, %Y"),
+                        'due_time': (f'{block_date.strftime("%H:%M %p")} GMT{timezone_offset}'
+                                     if block['assignment_type'] else '')
+                    })
+                else:
+                    course_date_list[0]['assignment_count'] += 1
```

Comment:
can we define a function like `prepare_date_block` that will receive `block_date` and `timezone` and return a prepared date block?